chore(vrl): CSV enrichment RFC (#8400)
timberio/vector: rfcs/2021-07-20-8288-csv-enrichment.md

# RFC 8288 - 2021-07-20 - CSV Enrichment

Vector needs to allow users to enrich the events flowing through the topology

> using through the topology using a CSV file. This RFC proposes adding a Vrl

Reviewer: ```suggestion
using a CSV file. This RFC proposes adding a VRL
```

function will not handle any indexing of this data. Above a certain size
indexing will provide substantial performance benefits, so it is something that
we should definitely plan for implementing at a future stage. The RFC will 
briefly discuss this, but will not go into any specific indexing strategies.

## Motivation

Users would like to enrich events flowing through Vector with extra data 
provided by a CSV file. 

## Doc-level Proposal

Two remap functions:

### `lookup_csv_row`

This function will look up a single row within the dataset. If a single row is 
found that data is returned as an object, otherwise this function will error. 

### `lookup_csv_rows` 

This function returns the results as an array of objects. If no row is found an
empty array is returned. Multiple results just result in multiple rows in the 
array.

### Parameters 

*filename*

The filename is the path to the filename of the csv file to lookup. This
filename has to be a static string (cannot be changed at runtime) so that 
Vector can:

1. Validate that the file exists.
2. Load the file into memory.

The filename will be relative to the current working directory that Vector is

> running in.

Reviewer: So, I'm not convinced this is the right strategy to take.

My preference would be as follows:

1. Allow to define a list of paths to which the VRL compiler/runtime has access in Vector's TOML file
2. Pass those to the VRL compiler and add them to the VRL context object
3. Introduce a new `p'...'` static type that represents a relative file location (i.e. it's a compile-time error to provide an absolute path or an invalid path string), and matches the first file it finds within the provided list of root paths defined in the Vector config.

Here's an example:

```toml
[transforms.remap]
type = "remap"
whitelist_paths = ["/opt/mycsvs"]
source = """
	lookup_csv_row(.data, p'my.csv')
"""
```

This has the following advantages:

1. It separates path location access from VRL source, meaning your VRL source easier to maintain (especially when we introduce loading VRL from a separate file (#6403).
2. It allows us to provide very specific compiler errors for path errors (e.g. invalid path syntax, non-existing files, read-access, etc)
3. Eventually (maybe, if need be), we can lift the `whitelist_paths` (or whatever we'll name it) up into a top-level Vector configuration option for other components to use.

Author: I'm wondering if it might be even better to move the paths *and* files out of the transform altogether.  @zsherman had a sample config example along the lines of:

```
global.csvfile = "/data.csv"

[transforms.enrich]
  source = '''
     . |= lookup_csv_row(global.csvfile, { "name": .name })
  '''
```

The advantage of this, in addition to the benefits you have mentioned, it also allows transforms to share the one data file. It's certainly possible that a given configuration may want multiple different transforms all enriching from the same datafile.

The CSV file must start with a title row that gives a name to each of the 

> columns provided in the file.

Reviewer: Can't we support non-titles by naming the columns according to their index value? E.g.

```csv
foo,bar,baz
```

```json
{
	"0": "foo",
	"1": "bar",
	"2": "baz"
}
```

Reviewer 2: 👍 I like that. We would probably want a "first row is headers" parameter to tell the function when there are headers.

*key*

The CSV file may be encrypted. This parameter specifies a key to use to decrypt
the file.

Is it safe to assume/require the file to be encrypted using AES, or should we

> support other algorithms as well?

Reviewer: This is a tricky one.

I think it's fine to assume this for a v1, but this doesn't scale well, especially given CSV is but one way to store data, and we might want to introduce other functions in the future that are basically the same, but read a JSON file, or any other data file.

This _might_ even be an even stronger reason to move file access outside VRL.

For example, you could do something like:

```toml
[transforms.remap]
type = "remap"
whitelist_paths = [
	{ name = "mycsv", decrypt = { ... }, path = "/opt/mycsvs/my.csv" }
]
source = """
	lookup_csv_row(.data, p',mycsv')
"""
```

Basically, this would be an extension to what I mentioned previously:

- You _have_ to specify individual files you want to access in Vector's config, it reads the files at boot time, and provides a reference to the content to VRL's compiler when it is initialized.
- The name you give it has to be unique, and is used to reference the file inside VRL
- VRL has a `p'<unique name>'` literal that maps the name to the data belonging to that file
- Because VRL receives a reference to the decrypted file contents, we don't have to worry about any function arguments that have to do with decryption.

Reviewer 3: or even a global setting like:

```
cache.csv.custom_data = "/data.csv"
```

Reviewer 2: Agreed, the decryption bit especially seems like it would motivate a separate "load" step.

If we go that route, we could use `f'mycsv'` to indicate it is a file (rather than `p` for "path").

Another way this could be modeled is like:

```
csv = open_csv("foo.csv", ...)
lookup_csv_row(csv, ...)
```

But this would be a bit of a first in VRL to create objects that are used later so I think I prefer the separate configuration in the `remap` transform. The alternative also gives the impression that the file is loaded every time.

However, modeling it as `remap` configuration also means that this VRL feature would be `remap` specific and not usable in other places VRL is used (like `filter`, `route` and unit tests) unless we add this configuration there too.

Reviewer: > However, modeling it as `remap` configuration also means that this VRL feature would be `remap` specific and not usable in other places VRL is used (like `filter`, `route` and unit tests) unless we add this configuration there too.

Correct, which is why I think we want to lift this into a top-level domain at some point. For now, we could add a component option that you can easily copy/paste to other components, similar to our existing `TlsConfig`, `BatchConfig` and the like.

We would also update the REPL to allow repeated use of arguments such as `--file mycsv=/foo/my.csv` to act similar to Vector's config field. For now, I'm assuming we can skip decrypting files in the REPL, but I imagine we can add `--file-decrypt-aes mycsv=$KEY` to specifically force decrypting provided files.

> If we go that route, we could use `f'mycsv'` to indicate it is a file (rather than `p` for "path").

Agreed, that does seem better if we go with the "reference the exact file" route.

Author: I'm really liking the top level configuration option.

Another consideration is that we need to allow for refreshing the file contents. It would not be a good solution to force people to have to restart Vector if they want to update the file. If we make it a top level option it would fit better with either allowing it to refresh the file on a SIGHUP, or perhaps watching the file for modification time changes, or even inotify.

When we add additional data sources, some of those may need to fit into the async ecosystem. Keeping the data source outside or VRL alleviates VRL from having to fit into that, keeping the core functionality simpler.

Downside, it does make the upfront development of this feature more complicated.

Reviewer 4: I'd note that @blt was recently talking about the removal of config watching, I believe in the pipelines RFC

Reviewer 2: Interesting, yeah, I'm down to entertain that. For now, we should probably aim for this to integrate with the existing `SIGHUP` reloading.

Reviewer 5: If I may jump in here 😄 . I do not think we should concern ourselves with encryption or reloading at this stage. Let's just get a simple static CSV file working, we can follow up with those enhancements.

Reviewer: > I do not think we should concern ourselves with encryption

I was somewhat believing that encryption was a requirement put forward by the people who requested this function. If not, then I fully agree that we should move encryption to a follow-up RFC whenever that becomes relevant.

Reviewer 5: Yes, that's my bad. I wasn't aware this was a requirement in the brief. Even so, I think we'd benefit from deferring it on this first version and focusing on the core functionality. I agree that we should consider it for our design, ensuring that we can accomodate it easily, but not solve it right now.

Reviewer: The proposal I made w.r.t. moving file reading outside VRL is almost entirely driven by the brief's requirement of decryption support.

The problem is that this is going to affect every file-based function we're adding, and I fully expect there to be more, other than the CSV one this RFC proposes. This is in addition to the complication of having to special-case path strings in the proposal you made w.r.t. having fully qualified paths be passed into the function as an argument.

Whatever way you slice or dice it, requiring local disk access _and_ having to support file decryption is going to add complications, either inside VRL, or out. My reasoning for moving it out of VRL is because it allows us more flexibility on how we implement it, and reduces the complicated matrix of parameters we're going to have to accept going forward (think: providing the file path itself, in addition to supporting a myriad of encryption schemes and their different options (which was well explained by @bruceg  in https://github.com/timberio/vector/issues/4870#issuecomment-721997201).

I think it's worth fleshing out both ways of solving this (entirely inside VRL, or through Vector's config) and list the pro's and con's, before we make a final decision on this.

*criteria*

`criteria` is a single level, key/value object that specifies the fields and
values to lookup. The fields must all match (AND) for the row to be returned.

*case_insensitive*

By default the search will be case sensitive, but this can be changed by 
passing `true` to this parameter.

## Internal Proposal

- Describe your change as if you were presenting it to the Vector team.
- Use lists, examples, and code blocks for efficient reading.
- Be specific!

The majority of the changes will be provided by adding two new functions within
`vrl/stdlib` and some shared lookup code.

Since the CSV file is loaded at bootime, it is not essential that it is loaded 
using async. This makes things much simpler as currently the VRL compilation 
process is entirely synchronous. No code outside of the new function in the
vrl stdlib should need changing.

The entire data file will be loaded into memory, so all lookups will be 
performed in memory.

> ### Indexing 

Reviewer 2: At a certain point of complexity in requirements, we may want to push people to sqlite too.

We _could_ even use an in-memory sqlite database and populate it from the CSV at start time. Users could configure indexes if we opt for configuration like what Jean has above:

```
files.myothercsv = {  decrypt = { ... }, path = "/opt/mycsvs/other.csv", index = ["foo", "bar", ["baz", "boop"]}
```

Although the first version is not going to be doing any indexing into the data
it is worth bearing it in mind as we will most likely need to add this in due
course.

In order to perform the indexing VRL needs to know which fields to index. The 
criteria is being passed in as an object. If the type def for that object is
known at runtime, we can extract the fields from this. 

If the type def isn't known we could 

- raise an error 
- perform the search unindexed
- create the index at runtime - generally the shape of this object will not 
  change over time, so most likely the cost of creating the index would only
  need to be paid on the first event.

The approach we take here should be decided before the first release so we don't
have to introduce a breaking change further down the line.

Actual indexing strategies can be decided later.

## Rationale

There is significant customer demand for this feature.

## Drawbacks

1. There are a number of issues with parsing CSV files in particular around

>    handling delimeters and separators.

Reviewer: I was thinking the same. I'm pretty sure we'll need a function argument to allow defining the delimiters because `,` and `;` are both very common.

Reviewer 2: Yeah, we should probably support the same parameters that `parse_csv` does.

2. Performance. Whilst we can ensure all the IO is performed at boot time,
   searching through that data could still be quite expensive.
3. Memory use. Since the data file is loaded into memory, the data will use up 
   some memory. If the file is large enough that could have an impact should

>    Vector be running in a constrained environment.

Reviewer 2: Yeah, we should make sure to call this out in the docs for this feature.

## Alternatives

### Use a predicate for the search

Instead of using an object to specify the search criteria we could allow the
user to specify a predicate to determine the row to use for enrichment.

```
lookup_csv_row("/path/to/file.csv", |row| row.some_key = .some_field)
```

> Note the ability to specify closures this is not yet available in VRL.

Reviewer: I would not favour this right now, at least not until we've fleshed out #6031 and #3740.

Reviewer 2: Yeah, agreed, it seems like this feature will ship first.

# RFC 8288 - 2021-07-20 - CSV Enrichment

Vector needs to allow users to enrich the events flowing through the topology
using a CSV file. This RFC proposes adding a VRL function that can perform 
this lookup.

## Scope

> The RFC will cover a simple VRL function that will lookup a single row from a
> CSV file. 

Reviewer 5: ```suggestion
The RFC will cover a simple VRL function that will lookup a single row from a
CSV file using a set of conditions. This allows the user to map the data into
the event using the full power of VRL.
```

> The lookup will be a simple sequential scan. The first version of this
> function will not handle any indexing of this data. Above a certain size
> indexing will provide substantial performance benefits, so it is something that
> we should definitely plan for implementing at a future stage. The RFC will 
> briefly discuss this, but will not go into any specific indexing strategies.

Reviewer 5: ```suggestion
For MVP purposes, the lookup will be a simple sequential scan. The first version
of this function will not handle any indexing of this data. Above a certain size,
indexing will provide substantial performance benefits, it is something that
we should definitely consider at a future stage. Therefore, it is out of scope for
this RFC.
```

## Motivation

Users would like to enrich events flowing through Vector with extra data 
provided by a CSV file. 

## Doc-level Proposal

### Resources

To represent the CSV file we have a new top level configuration option.

```

> [resources.csv_file]

Reviewer 2: We don't have to outline all of the metrics here, but I think it'd be useful to think about instrumenting these resources so that users can get an idea of things like file i/o.

Author: Not sure I fully understand. Do you mean emit metrics for filesize, load time etc?

Reviewer 2: Ah, true, I forgot we were loading into memory at start so there wouldn't really be on-going metrics for it outside of, say, how long a given `lookup_csv_row` call takes which would be a separate concern (but probably interesting to users).

  type = "csv"
  path = "\path_to_csv"
  key = "{KEY}"

>   delimeter = ","

Reviewer 2: We'll probably want to think about how this integrates with `vector validate`

>   delimeter = ","

Reviewer 5: I'm not a fan of this. It feels premature to define an abstract concept like "resources" just for CSV data. A simpler approach would be to supply these details as arguments to the VRL function. We can still load the file as a "resource" during compilation time, but the user isn't required to jump through those hoops.

```

The fields available for this section are:

*type* 

The type of data that the resource represents. Currently only a csv file is 
supported, but this may expand in future to handle other types of files and
database connections.

*path*

The path to the csv file, either an absolute path or one relative to the current
working directory.

*key*

The CSV file may be encrypted. This parameter specifies a key to use to decrypt
the file.

Is it safe to assume/require the file to be encrypted using AES, or should we
support other algorithms as well?

*delimeter*

The delimeter used in the csv file to separate fields. Defaults to `","`.

*title_row*

Defaults to true. If set it assumes the first row in the csv file contains 
column names. If false columns are named according to their numerical index.

### Vrl functions

Two remap functions:

> #### `lookup_csv_row`

Reviewer 5: The `lookup_*` prefix is a new convention and I don't think it's the best term. How about the following prefixes?

* `find_*` - finds a _single_ record
* `search_*` - finds _multiple_ records

This will set precedence for future functions that read from other sources, like a database.

This function will look up a single row within the dataset. If a single row is 
found that data is returned as an object, otherwise this function will error. 

#### `lookup_csv_rows` 

This function returns the results as an array of objects. If no row is found an
empty array is returned. Multiple results just result in multiple rows in the 

> array.

Reviewer 5: I don't think we need this to start. We haven't come across a use case that requires multiple rows.

cc @zsherman for confirmation.

#### Parameters 

*resource*

The name of the resource to lookup. This must point to a resource specified

> in the config file eg `resources.csv_file`.

Reviewer 5: Again, I dislike this. I'd much rather the user just specify the file path.

*criteria*

`criteria` is a single level, key/value object that specifies the fields and

> values to lookup. The fields must all match (AND) for the row to be returned.

Reviewer 5: I think `condition` is a better name for this. We should also consider the syntax, users will undoubtedly want more powerful matching conditions in the future (like SQL), but I agree that we should not solve that in the first version.

*case_insensitive*

By default the search will be case sensitive, but this can be changed by 

> passing `true` to this parameter.

Reviewer 5: Eh. If we're going to support this we should consider a different matching syntax like I mentioned above.

## Internal Proposal

We will need to add a new component type to Vector, call it `Resource`. On 
loading the Vector config these instances will be created and will load the 

> data that they are pointing to. 

Reviewer 5: Again, I disagree with exposing this concept to users.

   some memory. If the file is large enough that could have an impact should
   Vector be running in a constrained environment.

## Alternatives

### Use a predicate for the search

Instead of using an object to specify the search criteria we could allow the
user to specify a predicate to determine the row to use for enrichment.

```
lookup_csv_row("/path/to/file.csv", |row| row.some_key = .some_field)
```

Note the ability to specify closures this is not yet available in VRL.

The advantage of this is it allows the user to specify a more complex
set of criteria beyond simple equality. The downside is that it would be very
hard use indexes to ensure the lookup remains performant.

There is nothing that would prevent us from providing both options.

### Provide other data sources

This RFC proposes using a CSV file as a datasource. It time we may also need to
source the data from a JSON file, SQL database or Http source.

Using SQL and Http in such a way that would require VRL to perform a lookup
each event would significantly impact performance and would also require a 
fairly significant modification to VRL to allow VRL functions to run 
asynchronously.

## Outstanding Questions

> - What approach should we taken around determining the indexed fields?

Author: Yes, I agree.  It says at the top:

> The first version of this function will not handle any indexing of this data.

But I also think that since it is incredibly likely that we will have to add this, we should consider it in the design from the start to avoid backing us into a corner.

Configuration on the `resources` block is a possibility, and would be essential if we allowed remap to do searches on dynamic fields where we know what they are searching on at boot time.

> - Can we assume that if the file is encrypted it will use AES?

Author: I think just the key is sufficient, but will need to double check..

> - Should `title_row` default to true or false?

Author: Yeah, makes sense when the default isn't obvious.

# RFC 8288 - 2021-07-20 - CSV Enrichment

Vector needs to allow users to enrich the events flowing through the topology
using a CSV file. This RFC proposes adding a VRL function that can perform
this lookup.

## Scope

The RFC will cover a simple VRL function that will lookup a single row from a
CSV file using a set of conditions. This allows the user to map the data into
the event using the full power of VRL.

For MVP purposes, lookup will be a simple sequential scan. The first version
of this function will not handle any indexing of this data. Above a certain size
indexing will provide substantial performance benefits, so it is something that
we should definitely consider at a future stage. Therefore, it is out of scope
for this RFC.

Encryption will not be covered. This is very much a requirement to be considered
at a future stage, but is out of scope for this RFC.

Likewise, reloading the enrichment data is likely to be needed, but will not be
covered here. For the first implementation of this RFC, reloading the data will
require Vector to be restarted.

## Motivation

Users would like to enrich events flowing through Vector with extra data
provided by a CSV file.

> ## Doc-level Proposal

Reviewer 5: ```suggestion
## User Experience
```

> ### Resources

Reviewer 5: After thinking about @lukesteensen's proposal (join transform) and lightly discussing how the "resource" concept will evolve, I think a much better name for this is "table". It's just a table of data that we can build however we want (front loading, streaming, etc). To start, the CSV table will be all strings, but we can follow up with a schema enhancement if we desire. I think this naming is more intuitive since it aligns with database terminology and is a little more specific to it's purpose.

Author: Yes, that's a great idea. There was a vagueness in the name Resource that could cause people to think it's for all sorts of other things. Plus there is already the concept of Resource in the [source code](https://github.com/timberio/vector/blob/master/src/config/mod.rs#L466-L471) that is completely unrelated, so it makes it easier there too.

access to. Admin will likely want to restrict access to resources to approved
pipelines, especially if sensitive information is contained.

The resources will need to be integrated with `vector validate`, which would
ensure the resource exists and is correctly formatted.

### Schema

For the CSV Resource all columns will be considered to be Strings. Since
Resources are loaded before VRL compilation it will be possible to ensure that
Vrl doesn't search on columns that do not exist within the datafile. Searching
on a column that doesn't exist can prevent Vector from loading.

If (when these features are implemented) the user attempts to reload a Resource
and a column that has been Indexed no longer exists, this should prevent the
file from being loaded. Vector will continue to use the currently loaded data.

These restrictions apply only to columns that VRL is searching on. If the user
attempts to use a field that has not been returned from the enrichment, this
will be possible, but the value will be Null. This is largely due to the way
that VRL works with paths into Objects.

### Vrl functions

Two remap functions:

> #### `find_row`

Reviewer 5: To be clear, this is a generic function that will act on _all_ table types, right? If so, I'd prefer that we rename this to `find_table_row` (see [this comment](https://github.com/timberio/vector/pull/8400/files#r678291702) for naming) .

Author: > To be clear, this is a generic function that will act on all table types, right?

Correct.

Reviewer 5: Awesome, I like that a lot.

This function will look up a single row within the dataset. If a single row is
found that data is returned as an object, otherwise this function will error.

A metric will be emitted to indicate the lookup time.

> #### `search_rows`

Reviewer 5: Same as above, `search_table_rows`.

empty array is returned. Multiple results just result in multiple rows in the
array.

A metric will be emitted to indicate the lookup time.

#### Parameters

*resource*

The name of the resource to lookup. This must point to a resource specified
in the config file eg `resources.csv_file`.

*condition*

`condition` is a single level, key/value object that specifies the fields and
values to lookup. The fields must all match (AND) for the row to be returned.

*case_sensitive*

By default the search will be case insensitive, but this can be changed by
passing `true` to this parameter.

### Example config

```toml
[resources.csv]
    type = "csv"
    file = "/path/to/csv.csv"
    delimiter = ","

[sources.datadog_logs]
    type = "datadog_logs"
    address = "0.0.0.0:80"

[transforms.simple_enrich]
    type = "remap"
    inputs = ["datadog_logs"]
    source = '''
        . = parse_json!(.message)

        result, err = find_row(
            resources.csv,
            { "license_plate": .license }
        )

        if is_null(err) {
            .first_name = result.first_name
            .last_name = result.last_name
        }
   '''
```

> ## Internal Proposal

Reviewer 5: ```suggestion
## Implementation
```

We will need to add a new component type to Vector, call it `Resource`. On
loading the Vector config these instances will be created and will load the
data that they are pointing to.

> The entire data file will be loaded into memory, so all lookups will be

Reviewer 5: ```suggestion
The entire data file will be loaded into memory upon starting Vector, so all lookups will be
```

performed in memory. A `Resource` will need to provide threadsafe,
readonly access to the data that it loads.

VRL will need to maintain the concept of `Resource`. This can be created as
an additional element to the `vrl::Value` type. On compilation, the available
resources can be added to the Variable type definitions. This ensures that
during compilation the functions will access valid resources.

> ### Indexing

Reviewer 5: Actually, you note in the "scope" section that this is out of scope. So should we remove this section altogether?

Author: I think it is worth keeping this in. I would like to keep the restriction to ensure we know the search fields at compile time for the first version. It wouldn't strictly be necessary without indexing, but since we are anticipating indexing this helps to explain the restriction.

Although the first version is not going to be doing any indexing into the data
it is worth bearing it in mind as we will most likely need to add this in due
course.

In order to perform the indexing VRL needs to know which fields to index. The
criteria is being passed in as an object. If the type def for that object is
known at runtime, we can extract the fields from this.

If the type def isn't known we could

- raise an error
- perform the search unindexed
- create the index at runtime - generally the shape of this object will not
  change over time, so most likely the cost of creating the index would only
  need to be paid on the first event.

The approach we take here should be decided before the first release so we don't

> have to introduce a breaking change further down the line.

Reviewer 5: Can we choose one and resolve the comment here?

Author: Yes, I vote "raise an error". 

There is no requirement for us to perform dynamic searches at present, so it would be a case of choosing the simplest option that works. Plus if we change it down the line it will be relaxing restrictions rather than increasing them, thus preserving backward compatibility.

Reviewer 5: Works for me. Let's update with that then.

Since the resource file is loaded outside of VRL whilst determining which keys
need indexing occurs inside VRL, we will need a way for VRL to indicate to the
resource which indexes need building.

Actual indexing strategies can be decided later.

## Rationale

> There is significant customer demand for this feature.

Reviewer 5: To be clear, this section rationalizes your proposal, including the implementation, not just the business-level feature. Ex: Why is your approach better than the alternatives?

## Drawbacks

1. There are a number of issues with parsing CSV files in particular around

>    handling delimiters and separators.

Reviewer 5: Why exactly is this a drawback?

Author: It's a drawback with CSV over say JSON which is a more precise format.

  before the event data so that all events pass through enriched.

The join transform can be configured to specify the keys to use to join the two
streams.

The advantages of this is that it allows us to reuse the sources that we already
have. The user isn't limited to the data source that we provide as a Resource.

Downsides are that there is a data race at the start. The transform needs some
way of determining that all the necessary data has been loaded from the
Enrichment source before it starts to accept data from the Event source. This
can be tricky to get right.

Another complication could be determining a way to update the Enrichment data.
New records could simply be streamed in by appending to the file, but modified
data could only be updated if we ensured the data also had some form of primary
key.

Because of these, it is felt the Join transform dosn't fit the exact problem of
enrichment we are currently solving, however there are many other scenarios
where this could be useful. In future, we can create a new Resource type that
can accept an input from a Vector source which would provide all the benefits of

> the Join transform.

Reviewer 5: Thanks for adding this, it makes sense.

# RFC 8288 - 2021-07-20 - CSV Enrichment

Vector needs to allow users to enrich the events flowing through the topology
using a CSV file. This RFC proposes adding a VRL function that can perform
this lookup.

## Scope

The RFC will cover a simple VRL function that will lookup a single row from a
CSV file using a set of conditions. This allows the user to map the data into
the event using the full power of VRL.

> For MVP purposes, lookup will be a simple sequential scan. The first version

Reviewer 2: Agreed. I think this will be a very fast follow-up.

> For MVP purposes, lookup will be a simple sequential scan. The first version

Author: > indexing will happen almost immediately after the initial implementation and make sure we implement with that in mind.

This is my understanding. I wouldn't consider the feature to be complete and ready for release without indexing.

of this function will not handle any indexing of this data. Above a certain size
indexing will provide substantial performance benefits, so it is something that
we should definitely consider at a future stage. Therefore, it is out of scope
for this RFC.

Encryption will not be covered. This is very much a requirement to be considered
at a future stage, but is out of scope for this RFC.

Likewise, reloading the enrichment data is likely to be needed, but will not be
covered here. For the first implementation of this RFC, reloading the data will
require Vector to be restarted.

## Motivation

Users would like to enrich events flowing through Vector with extra data
provided by a CSV file.

## User Experience

> ### Tables

Reviewer 2: Agreed, I don't think `table` will map as well when we introduce other data sources.

> ### Tables

Reviewer 5: Eh, I'm not sure. It's hard to say until we work through that data source type. I can see this term fitting well with certain design solutions.

> ### Tables

Author: Am open to suggestions! How about DataFrame?

> ### Tables

Reviewer 6: I'm trying to think of something that communicates "thing against which you can do lookups for enrichment" without implying a particular physical representation. In theory we could extend this concept to things like the EC2 metadata API, Postgres databases, Redis, etc.

We could maybe do something like `lookup`, but that clashes with the event-level idea of a lookup. It's a little long, but `enrichment_source` kinda captures it. Maybe `lookup_source`, `query_source`, `queryable`, ...?

To represent the CSV file we have a new top level configuration option.

```
[tables.csv_file]

>   type = "csv"

Author: I'm in two minds about this. One one hand there would be a lot of shared fields between say a json and a csv file table, but there is one field - `delimiter` which would apply only to csv. I'll change it to a `file` type for now, but that may be liable to change back..

>   type = "csv"

Reviewer 6: This might be something where we just go with the specific version for now and then make a call about whether or not to generalize before the feature is actually released.

>   type = "csv"

Reviewer 2: Yeah, that's a good point. There will be options specific to the file type. We could probably still model that like:

```toml
type = "file"
encoding.type= "csv"
encoding.csv.delimiter = ","
encoding.csv.header_row = true
```

Where the CSV encoding options would only apply if `encoding.type = "csv"`.

file from being loaded. Vector will continue to use the currently loaded data.

These restrictions apply only to columns that VRL is searching on. If the user
attempts to use a field that has not been returned from the enrichment, this
will be possible, but the value will be Null. This is largely due to the way
that VRL works with paths into Objects.

### Vrl functions

Two remap functions:

#### `find_table_row`

This function will look up a single row within the table dataset. If a single
row is found that data is returned as an object, otherwise this function will
error.

A metric will be emitted to indicate the lookup time.

#### `search_table_rows`

This function returns the results as an array of objects. If no row is found an
empty array is returned. Multiple results just result in multiple rows in the
array.

A metric will be emitted to indicate the lookup time.

#### Parameters

*table*

The name of the table to lookup. This must point to a table specified in the
config file eg `tables.csv_file`. Both functions are generic over all Table
types.

*condition*

`condition` is a single level, key/value object that specifies the fields and

> values to lookup. The fields must all match (AND) for the row to be returned.

Reviewer 6: Could we use named args here instead of an object? E.g. `find_table_row(tables.csv_file, license_plate: .license)`. That seems a little more natural and emphasizes that the lookup key (i.e. `license_plate`) is static.

I'd also lean toward starting out with a single key lookup for this. If there's user demand, we can add multiple keys later.

Reviewer 2: I'm cool with starting with a single key but the user that motivated this feature does specifically have a need for composite keys.

For named arguments, we don't currently have support dynamic keys (and hesitate to introduce that as part of this feature). I think we could do `find_table_row(tables.csv_file, { "license_plate": .license } )` though.

Reviewer 6: Ah, good to know. A single composite key lookup is still quite a bit simpler than supporting multiple lookup keys. Basically, the hierarchy of complexity goes:

1. Single scalar key:
```
find_table_row(..., { "license": .license })
```
2. Single composite key:
```
find_table_row(..., { "license": .license, "name": .name })
```
3. Multiple keys:
```
find_table_row(..., { "license": .license })
find_table_row(..., { "name": .name })
```

I'm just suggesting that we only go as high as we need to for the first iteration to be useful.

Author: That doesn't fit so well with the way VRL has been designed, so would require a more extensive change to the language. It's certainly doable if it is something you feel strongly about, but if not I would prefer to stick with an object for now.

We already know there is user demand for multiple keys, so do need to make this work.

In terms of indexing, since all lookups are AND, it should be a fairly straightforward operation to  index based on the combined hash of the fields to lookup. This would get more complicated should more complicated searches be required, but we can defer that to a later date should that happen.

*case_sensitive*

By default the search will be case insensitive, but this can be changed by
passing `true` to this parameter.

### Example config

```toml
[tables.csv_file]
    type = "csv"
    file = "/path/to/csv.csv"
    delimiter = ","

[sources.datadog_logs]
    type = "datadog_logs"
    address = "0.0.0.0:80"

[transforms.simple_enrich]
    type = "remap"
    inputs = ["datadog_logs"]
    source = '''
        . = parse_json!(.message)

        result, err = find_table_row(

>             tables.csv_file,

Reviewer 6: Do we expect to be able to look up things other than tables? If not, we could maybe shorted this to just `find_table_row(csv_file, ...)`.

Author: 'tables' is just a standard VRL variable that will just be injected into the script. I think it would be better to namespace them all under a single variable `tables` to avoid conflicting with other variables the user may want to use.

Reviewer: I'm not fond of magic variables (can people overwrite them, if not, we have to add some variable names to our list of reserved identifiers).

I do wonder how this intersects with event metadata and the desire to expose that information in VRL at some point.

I don't have a solution at hand, but either we reserve `tables` (and `metadata`) as reserved identifiers, or we find another way to allow operators to access special datasets that are separate from event fields and regular variables.

Author: People could overwrite `tables`, but if the script tried to use it in a call to `find_table_row` it would no longer compile since it would then be the wrong type. The type system should be sufficient to keep everything safe here.

would be able to take two inputs:

- The Event source
  The data containing the events to be enriched.

- The Enrichment source
  This source would contain the enrichment data. It needs to be loaded in
  before the event data so that all events pass through enriched.

The join transform can be configured to specify the keys to use to join the two
streams.

The advantages of this is that it allows us to reuse the sources that we already
have. The user isn't limited to the data source that we provide as a Resource.

Downsides are that there is a data race at the start. The transform needs some
way of determining that all the necessary data has been loaded from the
Enrichment source before it starts to accept data from the Event source. This
can be tricky to get right.

Another complication could be determining a way to update the Enrichment data.
New records could simply be streamed in by appending to the file, but modified
data could only be updated if we ensured the data also had some form of primary
key.

Because of these, it is felt the Join transform dosn't fit the exact problem of
enrichment we are currently solving, however there are many other scenarios

> where this could be useful. In future, we can create a new Resource type that

Reviewer 6: ```suggestion
where this could be useful. In future, we can create a new Table type that
```

- What approach should we taken around determining the indexed fields?
- Can we assume that if the file is encrypted it will use AES?
- Should `title_row` default to true or false?
- Does the parameter `case_insensitive` indicate that perhaps does a simple 
  key/value search is not sufficient.

## Plan Of Attack

- [ ] Add support for resources. Any resource sections will load the data at 
      boot time and will prevent Vector from starting up if the resource file

>       cannot be found or is incorrectly formatted.

Reviewer: This alone seems to be worthy of a separate RFC, I'm not sure what you, @jszwedko and @binarylogic agreed on, but I would advocate for one RFC that specifically focusses on the addition of the "resource" type, which also describes (not in exact details, but exact enough that we _know_ it's doable) how we're going to leverage the resource type for other types of resources (database, external URL), and how we're going to handle encrypted resources.

Once that RFC lands, this RFC can simply reference that one, and expand on it by explaining how the resource is passed to VRL, how you reference your resource, and how the actual function is going to work.

Reviewer 5: I agree, although it feels premature at this point. I'd much rather wait until we introduce another resource type to consider this. Hence why I wasn't onboard with resources to begin with.

I think it's fine to introduce for now, as long as it's CSV specific.

Reviewer: > I think it's fine to introduce for now, as long as it's CSV specific.

Just to clarify, that means you're okay with making breaking changes if we find out we need to tweak the layout of the resource configuration if it turns out to be inflexible w.r.t. other resource types? Or do you mean we should settle that in this RFC and make sure we allow forward compatibility with resources such as databases, remote URLs, etc?

Reviewer 5: Sure, to be more specific, I want ot make sure we are not creating artificial restrictions when we decide to introduce another resource type, like a database. We should have complete free to integrate with that resource type in whatever way is most optimal. In regards to the options, I'd like to consider a patter that will allow for other resource types. Let's do our best for now. Just like anything, I'm ok breaking backward compatibility if it's worth it.
